packages/web3/src/entitlement.ts

```diff
@@ -767,12 +768,133 @@ export function createExternalNFTStruct(
+                }
+                if (this.decodedCheckOp.address === undefined) {
+                    throw new Error(`DecodedCheckOperation of type ${opStr} requires an address`)
+                }
+        }
+
+        // threshold check
+        switch (this.decodedCheckOp.type) {
+            case CheckOperationType.ERC1155: 
+            case CheckOperationType.ETH_BALANCE:
+            case CheckOperationType.ERC20:
+            case CheckOperationType.ERC721:
+                if (this.decodedCheckOp.threshold === undefined) {
+                    throw new Error(`DecodedCheckOperation of type ${opStr} requires a threshold`)
+                }
+        }
+
+        // tokenId check
+        if (
+            this.decodedCheckOp.type === CheckOperationType.ERC1155 &&
+            this.decodedCheckOp.tokenId === undefined
+        ) {
+            throw new Error(`DecodedCheckOperation of type ${opStr} requires a tokenId`)
+        }
+
+        // byte-encoded params check
+        if (
+            this.decodedCheckOp.type === CheckOperationType.ISENTITLED &&
+            this.decodedCheckOp.byteEncodedParams === undefined
+        ) {
+            throw new Error(`DecodedCheckOperation of type ${opStr} requires byteEncodedParams`)
+        }
+
+        switch (this.decodedCheckOp.type) {
+            case CheckOperationType.ERC20:
+            case CheckOperationType.ERC721:
+                return {
+                    type: this.decodedCheckOp.type,
+                    chainId: this.decodedCheckOp.chainId!,
+                    address: this.decodedCheckOp.address!,
+                    threshold: this.decodedCheckOp.threshold!
+                }
+
+            case CheckOperationType.ERC1155:
+                return {
+                    type: CheckOperationType.ERC1155,
+                    chainId: this.decodedCheckOp.chainId!,
+                    address: this.decodedCheckOp.address!,
+                    threshold: this.decodedCheckOp.threshold!,
+                    tokenId: this.decodedCheckOp.tokenId!,
+                }
+    
+            case CheckOperationType.ETH_BALANCE:
+                return {
+                    type: CheckOperationType.ETH_BALANCE,
+                    threshold: this.decodedCheckOp.threshold!,
+                }
+            
+            case CheckOperationType.ISENTITLED:
+                return {
+                    type: CheckOperationType.ISENTITLED,
+                    address: this.decodedCheckOp.address!,
+                    chainId: this.decodedCheckOp.chainId!,
+                    byteEncodedParams: this.decodedCheckOp.byteEncodedParams!
+                }
+
+            default:
+                throw new Error(`Check operation type ${opStr} unrecognized or not used in production`)
+        }
+    }
+}
+
 export type DecodedCheckOperation = {
     type: CheckOperationType
-    chainId: bigint
-    address: Address
+    chainId?: bigint
```

Comment:
chainId and address are actually not required for ETH_BALANCE checks, lets make these optional and ensure they exist when required by constructing DecodedCheckOperations with the builder in harmony.